052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,164 @@
+* This will allow the removal of the `GET` method, as the operator should not care about other fields. It should only know the desired state, and ask Kubernetes API to apply that, letting Kubernetes handle the merge.
+
+The following table shows changes in existing Kubernetes behaviour, with examples:
+| current value of field | desired value of field | field owner | requires update? | change in behaviour |
+| --------               | --------               | ------      | ---------------- | ------------------- |
+| x                      | x                      | Strimzi     | no               | yes                 |
+| x                      | y                      | Strimzi     | yes              | no                  |
+| x                      | not set                | Strimzi     | yes              | no                  |
+| x                      | not set                | Not Strimzi | no               | yes                 |
+| not set                | y                      | Strimzi     | yes              | no                  |
+| not set                | y                      | Not Strimzi | yes              | no                  |
+
+The key thing to take from this table is that currently Strimzi is trying to make changes to ALL fields (minus short `ignorableFields` list), once Strimzi has knowledge of the fields it owns - we don't want it to try and change the fields it doesn't own. Also Strimzi will be sending a PATCH call for every resource every reconciliation loop, so where no fields have changed it will be a NO-OP with little resource usage.
+
+---
+### Modifications to existing behaviours
+
+The only change to existing (non-Strimzi specific) behaviours is that `ResourceDiff` won't care about other fields it chooses not to care about. This does mean that another entity within the cluster can modify things without Strimzi's knowledge, but it shouldn't matter if the `metadata.managedFields` are chosen correctly.
+
+For users:
+
+* If you want to configure something for a single object (e.g. a single Kafka cluster), and it's in scope of the configuration, use that
+* If you want to configure something that's not the currently configurable fields, you can use other mechanisms to do so (Kyverno, other controllers), and Strimzi won't interfere
+* If you must configure something through some other method, and it's not directly being set in a different way by Strimzi, then Strimzi won't interfere and it'll "just work" (e.g. cattle annotations, pvc resizing annotations, metallb annotations. The user didn't make a choice about this other than the "I want to run Strimzi on this cluster that uses those technologies")
```

Comment:
I could follow this section, but I think it would be clearer and more concrete to users if we provided a couple of examples.

Reply:
I'm not sure what other examples I could add to this that would enhance it?
It is just saying if you want to use `annotations` that interact with other operators (such as the ones listed above) then you can, and Strimzi won't interfere?

Reply:
I'm saying that the language used to explain it (e.g. "If you want to configure something for a single object (e.g. a single Kafka cluster), and it's in scope of the configuration, use that") is quite abstract. A casual reader might struggle to really understand this in concrete terms (e.g. in that sentence does 'Kafka' mean a `Kafka` CR? Does 'configuration' mean the `spec.kafka.configuration`?). I think a snippet of YAML showing an example, for the `Kafka` CR, of the first and second bullets would really help such readers.

Reply:
Will attempt to rewrite to make it easier for the reader/user to understand 😄

Reply:
TBH, I'm not sure I understand the example. Strimzi would not care about the any of the annotations on the Kafka CR. Maybe it should be more something like this?

----

* If you want to configure part of a Strimzi Custom Resource (CR) (e.g. `Kafka` CR `replicas`) and the field exists within the CR, then you should use that field
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    template:
      pod:
        metadata:
          annotations:
            custom-annotation: "..." # defined with use of CR, therefore "owned" by Strimzi
    ...
```
* If you want to configure part of a Strimzi CR that is not within the configurable fields (e.g. `annotations` used by another operator) or has a dynamic value, you can apply them in any manner you like and Strimzi won't interfere (unless you define them in the CR itself).
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-cluster-kafka-0
  annotations:
    policies.kyverno.io/last-applied-patches: "..." # applied from Kyverno operator, therefore ignored by Strimzi operator
    custom-annotation: "..." # defined with use of CR, therefore "owned" by Strimzi
spec:
  ...
```